File: lib/go/telepact/internal/generation/GenerateRandomMockCall.go
//|
//|  Copyright The Telepact Authors
//|
//|  Licensed under the Apache License, Version 2.0 (the "License");
//|  you may not use this file except in compliance with the License.
//|  You may obtain a copy of the License at
//|
//|  https://www.apache.org/licenses/LICENSE-2.0
//|
//|  Unless required by applicable law or agreed to in writing, software
//|  distributed under the License is distributed on an "AS IS" BASIS,
//|  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//|  See the License for the specific language governing permissions and
//|  limitations under the License.
//|

package generation

import (
	"sort"
	"strings"

	"github.com/telepact/telepact/lib/go/telepact/internal/types"
)

// GenerateRandomMockCall produces a pseudo-random mock call structure based on the supplied schema types.
func GenerateRandomMockCall(typesMap map[string]types.TType, ctx *GenerateContext) any {
	if ctx == nil || ctx.RandomGenerator == nil {
		return nil
	}

	functionNames := make([]string, 0)
	for key := range typesMap {
		if strings.HasPrefix(key, "fn.") && !strings.HasSuffix(key, ".->") && !strings.HasSuffix(key, "_") {
			functionNames = append(functionNames, key)
		}
	}

	sort.Strings(functionNames)
	if len(functionNames) == 0 {
		return nil
	}

	selectedFnName := functionNames[ctx.RandomGenerator.NextIntWithCeiling(len(functionNames))]

	typeEntry := typesMap[selectedFnName]
	unionType, ok := typeEntry.(*types.TUnion)
	if !ok {
		return nil
	}

	alwaysIncludeRequired := false
	return GenerateRandomUnion(nil, false, unionType.Tags, ctx.Copy(nil, nil, &alwaysIncludeRequired, nil, nil))
}

File: lib/py/telepact/internal/generation/GenerateRandomMockCall.py
#|
#|  Copyright The Telepact Authors
#|  SPDX-License-Identifier: Apache-2.0
#|

from typing import cast

from .GenerateContext import GenerateContext
from .GenerateRandomUnion import generate_random_union
from ..types.TType import TType


def generate_random_mock_call(types: dict[str, TType], ctx: GenerateContext) -> object:
    from ..types.TUnion import TUnion

    function_names = [
        key for key in types.keys()
        if key.startswith('fn.') and not key.endswith('.->') and not key.endswith('_')
    ]

    function_names.sort()

    selected_fn_name = function_names[ctx.random_generator.next_int_with_ceiling(
        len(function_names))]
    
    selected_fn = cast(TUnion, types[selected_fn_name])

    return generate_random_union(None, False, selected_fn.tags, ctx.copy(always_include_required_fields=False))
